clippy_lints/src/semicolon_block.rs

```diff
@@ -0,0 +1,123 @@
+    ///
+    /// For () returning expressions, check that the semicolon is outside the block.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// For consistency it's best to have the semicolon inside/outside the block. Either way is fine and this lint suggests outside the block.
+    /// Take a look at `semicolon_inside_block` for the other alternative.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// unsafe { f(x); }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// unsafe { f(x) };
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub SEMICOLON_OUTSIDE_BLOCK,
+    restriction,
+    "add a semicolon outside the block"
+}
+declare_lint_pass!(SemicolonBlock => [SEMICOLON_INSIDE_BLOCK, SEMICOLON_OUTSIDE_BLOCK]);
+
+impl LateLintPass<'_> for SemicolonBlock {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &Block<'_>) {
+        semicolon_inside_block(cx, block);
+        semicolon_outside_block(cx, block);
+    }
+}
+
+fn semicolon_inside_block(cx: &LateContext<'_>, block: &Block<'_>) {
+    if !block.span.from_expansion()
+    && let Some(tail) = block.expr
+    && let Some(block_expr @ Expr { kind: ExprKind::Block(_, _), ..}) = get_parent_expr_for_hir(cx, block.hir_id)
+    && let Some(Node::Stmt(Stmt { kind: StmtKind::Semi(_), span, .. })) = get_parent_node(cx.tcx, block_expr.hir_id)
+    {
+        let expr_snip = snippet_with_macro_callsite(cx, tail.span, "..");
+
+        let mut suggestion: String = snippet_with_macro_callsite(cx, block.span, "..").to_string();
+
+        if let Some((expr_offset, _)) = suggestion.rmatch_indices(&*expr_snip).next() {
+            suggestion.insert(expr_offset + expr_snip.len(), ';');
+        } else {
+            return;
+        }
+
+        span_lint_and_sugg(
+            cx,
+            SEMICOLON_INSIDE_BLOCK,
+            *span,
+            "consider moving the `;` inside the block for consistent formatting",
+            "put the `;` here",
+            suggestion,
+            Applicability::MaybeIncorrect,
+        );
+    }
+}
+
+fn semicolon_outside_block(cx: &LateContext<'_>, block: &Block<'_>) {
+    if !block.span.from_expansion()
+    && block.expr.is_none()
+    && let [.., Stmt { kind: StmtKind::Semi(expr), .. }] = block.stmts
+    && let Some(block_expr @ Expr { kind: ExprKind::Block(_, _), ..}) = get_parent_expr_for_hir(cx,block.hir_id)
+    && let Some(Node::Stmt(Stmt { kind: StmtKind::Expr(_), .. })) = get_parent_node(cx.tcx, block_expr.hir_id) {
+        let expr_snip = snippet_with_macro_callsite(cx, expr.span, "..");
+
+        let mut suggestion: String = snippet_with_macro_callsite(cx, block.span, "..").to_string();
+
+        if let Some((expr_offset, _)) = suggestion.rmatch_indices(&*expr_snip).next()
+        && let Some(semi_offset) = suggestion[expr_offset + expr_snip.len()..].find(';') {
+            suggestion.remove(expr_offset +  expr_snip.len() + semi_offset);
+        } else {
+            return;
+        }
```

Comment:
I am not sure if the `find` is necessary since in theory the semi should be immediately at the start of the slice, but I wasn't sure if that was always gonna be the case